python_scripts/load_arrow.py

```diff
@@ -1,13 +1,295 @@
+        same name.
+        """
+        col_name_index = self.column_names.index(column_name)
+        self.arrow_table = self.arrow_table.remove_column(col_name_index)
+        self.arrow_table = self.arrow_table.add_column(
+            col_name_index, column_name, converted
+        )
+
+    def replace_missing_values(self):
+        """
+        Replace null values in the arrow table with the value stata needs.
+        Dates/timestamps/category require additional conversion before loading to
+        stata, so null replacements are deferred till later.
+        """
+        for column_name, column_type in self.column_types.items():
+            if column_type in ["date", "timestamp", "category"]:
+                # Skip dates/timestamps/category
+                # We replace these after conversion to stata dates in `load_data`
+                continue
+            null_value = self.MISSING_VALUES[column_type]
+            converted = self.arrow_table[column_name].fill_null(null_value)
+            col_name_index = self.column_names.index(column_name)
+            self.arrow_table = self.arrow_table.remove_column(col_name_index)
+            self.arrow_table = self.arrow_table.add_column(
+                col_name_index, column_name, converted
+            )
+
+    def get_categorical_encodings(self, varnames):
+        """
+        Retrieve a mapping of labels: values for categorical vars.
+        Returns a dict where the variable names are the keys and the value
+        is a dictionary consisting of label : value key/value pairs
+        """
+        value_labels = {}
+        for varname in varnames:
+            first_chunk = self.arrow_table[varname].chunks[0]
+            # Get the unique values for this variable and numeric mapping
+            value_map = {
+                str(value): idx for idx, value in enumerate(first_chunk.dictionary)
+            }
+            value_labels[varname] = value_map
+
+        return value_labels
+
+    def make_vars(self):
+        """
+        Function used to create the Stata variables in memory.
+        This assumes that the values in name_and_type object are valid
+        Stata variable names.
+        :param name_and_type: A dictionary structured to contain the variable names
+        as the keys and the storage type of the variable as the value.
+        """
+        # Loop over the variable name / variable type mappings and add
+        # variables with the appropriate typing based on the vartype
+        for varname, vartype in self.column_types.items():
+            if vartype == "string":
+                sfi.Data.addVarStrL(varname)
+            elif vartype == "boolean":
+                sfi.Data.addVarByte(varname)
+            elif vartype == "int":
+                sfi.Data.addVarInt(varname)
+            elif vartype == "category":
+                sfi.Data.addVarLong(varname)
+            elif vartype in ["float", "date", "timestamp"]:
```

Comment:
dates and timestamps are floats?!

oh my...

Reply:
Dates and timestamps are numeric in stata (afaict we only have to deal with dates, so days since 1st Jan 1960 is what we need to convert to, but they could be seconds), and b/c of the int size limit, they need to be float type. I think. I'm still somewhat vague on this.